src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,42 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbOptional = userService.findByEmail(email);
+        if (userFromDbOptional.isEmpty()
+                || !userFromDbOptional.get().getPassword()
+                .equals(HashUtil.hashPassword(password, userFromDbOptional.get().getSalt()))) {
+            throw new AuthenticationException("Email or password are incorrect");
+        }
+        return userFromDbOptional.get();
+    }
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        if (email == null || email.isEmpty() || email.isBlank()) {
+            throw new RegistrationException("Email can't be empty");
+        }
+        if (userService.findByEmail(email).isPresent()) {
+            throw new RegistrationException(String.format("Email %s already used", email));
+        }
+        if (password == null || password.isEmpty() || password.isBlank()) {
+            throw new RegistrationException("Password can't be empty");
+        }
```

Comment:
better switch these conditional. You don’t  need to make a request to DB if the password  is empty